src/AppBundle/EventSubscriber/RateLimitSubscriber.php

```diff
@@ -0,0 +1,95 @@
+<?php
+/**
+ * This file contains only the RateLimitSubscriber class.
+ */
+
+namespace AppBundle\EventSubscriber;
+
+use Symfony\Component\HttpKernel\Exception\AccessDeniedHttpException;
+use Symfony\Component\DependencyInjection\Container;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpKernel\Event\FilterControllerEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+use Psr\Cache\CacheItemPoolInterface;
+use DateInterval;
+
+/**
+ * A RateLimitSubscriber checks to see if users are exceeding usage limitations.
+ */
+class RateLimitSubscriber implements EventSubscriberInterface
+{
+
+    /** @var Container The DI container. */
+    protected $container;
+
+    /**
+     * Save the container for later use.
+     * @param Container $container The DI container.
+     */
+    public function __construct(Container $container)
+    {
+        $this->container = $container;
+    }
+
+    /**
+     * Register our interest in the kernel.controller event.
+     * @return string[]
+     */
+    public static function getSubscribedEvents()
+    {
+        return [
+            KernelEvents::CONTROLLER => 'onKernelController',
+        ];
+    }
+
+    /**
+     * Check to see if the current tool is enabled.
+     * @param FilterControllerEvent $event The event.
+     * @throws AccessDeniedHttpException If rate limits have been exceeded.
+     */
+    public function onKernelController(FilterControllerEvent $event)
+    {
+        $rateLimit = $this->container->getParameter('app.rate_limit_count');
+        $rateDuration = $this->container->getParameter('app.rate_limit_time');
+
+        // Zero values indicate the rate limiting feature should be disabled.
+        if ($rateLimit === 0 || $rateDuration === 0) {
+            return;
+        }
+
+        $controller = $event->getController();
+        $loggedIn = (bool) $this->container->get('session')->get('logged_in_user');
+
+        // No rate limits on index pages or if they are logged in.
+        if ($controller[1] === 'indexAction' || $loggedIn) {
+            return;
+        }
+
+        $request = $event->getRequest();
+        $sessionId = $request->getSession()->getId();
+
+        // Remove reserved characters from URI
+        $uri = preg_replace('/[\{\}\(\)\/\\@:]/', '', $request->getUri());
+
+        $cacheKey = 'ratelimit.'.$sessionId.'.'.$uri;
+        $cache = $this->container->get('cache.app');
+        $cacheItem = $cache->getItem($cacheKey);
+
+        // Initial count, assuming it's their first request with this URI
+        $count = $cacheItem ? (int) $cache->getItem($cacheKey)->get() + 1 : 1;
```

Comment:
When will `$cacheItem` be false? Won't it always be a `CacheItem`? I think this should be `$cacheItem->isHit()` in the first part (and the `$cacheItem` doesn't have to be gotten again in the second part).